File: src/test/java/org/luxons/sevenwonders/game/data/serializers/ScienceProgressSerializerTest.java
package org.luxons.sevenwonders.game.data.serializers;

import org.junit.Before;
import org.junit.Test;
import org.luxons.sevenwonders.game.boards.ScienceType;
import org.luxons.sevenwonders.game.effects.ScienceProgress;
import org.luxons.sevenwonders.game.test.TestUtils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import static org.junit.Assert.*;

public class ScienceProgressSerializerTest {

    private static final String COMPASS_STR = "\"COMPASS\"";

    private static final String WHEEL_STR = "\"WHEEL\"";

    private static final String TABLET_STR = "\"TABLET\"";

    private static final String JOKER_STR = "\"any\"";

    private Gson gson;

    @Before
    public void setUp() {
        gson = new GsonBuilder().registerTypeAdapter(ScienceProgress.class, new ScienceProgressSerializer()).create();
    }

    @Test
    public void serialize_emptyToNull() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 0, 0, 0);
        String json = gson.toJson(progress);
        assertEquals("null", json);
    }

    @Test
    public void serialize_oneCompass() {
        ScienceProgress progress = TestUtils.createScienceProgress(1, 0, 0, 0);
        String json = gson.toJson(progress);
        assertEquals(COMPASS_STR, json);
    }

    @Test
    public void serialize_oneWheel() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 1, 0, 0);
        String json = gson.toJson(progress);
        assertEquals(WHEEL_STR, json);
    }

    @Test
    public void serialize_oneTablet() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 0, 1, 0);
        String json = gson.toJson(progress);
        assertEquals(TABLET_STR, json);
    }

    @Test
    public void serialize_oneJoker() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 0, 0, 1);
        String json = gson.toJson(progress);
        assertEquals(JOKER_STR, json);
    }

    @Test(expected = UnsupportedOperationException.class)
    public void serialize_failOnMultipleCompasses() {
        ScienceProgress progress = TestUtils.createScienceProgress(2, 0, 0, 0);
        gson.toJson(progress);
    }

    @Test(expected = UnsupportedOperationException.class)
    public void serialize_failOnMultipleWheels() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 2, 0, 0);
        gson.toJson(progress);
    }

    @Test(expected = UnsupportedOperationException.class)
    public void serialize_failOnMultipleTablets() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 0, 2, 0);
        gson.toJson(progress);
    }

    @Test(expected = UnsupportedOperationException.class)
    public void serialize_failOnMultipleJokers() {
        ScienceProgress progress = TestUtils.createScienceProgress(0, 0, 0, 2);
        gson.toJson(progress);
    }

    @Test(expected = UnsupportedOperationException.class)
    public void serialize_failOnMixedElements() {
        ScienceProgress progress = TestUtils.createScienceProgress(1, 1, 0, 0);
        gson.toJson(progress);
    }

    @Test(expected = IllegalArgumentException.class)
    public void deserialize_failOnEmptyString() {
        gson.fromJson("\"\"", ScienceProgress.class);
    }

    @Test(expected = IllegalArgumentException.class)
    public void deserialize_failOnGarbageString() {
        gson.fromJson("thisisgarbage", ScienceProgress.class);
    }

    @Test
    public void deserialize_compass() {
        ScienceProgress progress = gson.fromJson(COMPASS_STR, ScienceProgress.class);
        assertNotNull(progress.getScience());
        assertEquals(1, progress.getScience().getQuantity(ScienceType.COMPASS));
        assertEquals(0, progress.getScience().getQuantity(ScienceType.WHEEL));
        assertEquals(0, progress.getScience().getQuantity(ScienceType.TABLET));
        assertEquals(0, progress.getScience().getJokers());
    }

    @Test
    public void deserialize_wheel() {
        ScienceProgress progress = gson.fromJson(WHEEL_STR, ScienceProgress.class);
        assertNotNull(progress.getScience());
        assertEquals(0, progress.getScience().getQuantity(ScienceType.COMPASS));
        assertEquals(1, progress.getScience().getQuantity(ScienceType.WHEEL));
        assertEquals(0, progress.getScience().getQuantity(ScienceType.TABLET));
        assertEquals(0, progress.getScience().getJokers());
    }

    @Test
    public void deserialize_tablet() {
        ScienceProgress progress = gson.fromJson(TABLET_STR, ScienceProgress.class);
        assertNotNull(progress.getScience());
        assertEquals(0, progress.getScience().getQuantity(ScienceType.COMPASS));
        assertEquals(0, progress.getScience().getQuantity(ScienceType.WHEEL));
        assertEquals(1, progress.getScience().getQuantity(ScienceType.TABLET));
        assertEquals(0, progress.getScience().getJokers());
    }

    @Test
    public void deserialize_joker() {
        ScienceProgress progress = gson.fromJson(JOKER_STR, ScienceProgress.class);
        assertNotNull(progress.getScience());
        assertEquals(0, progress.getScience().getQuantity(ScienceType.COMPASS));
        assertEquals(0, progress.getScience().getQuantity(ScienceType.WHEEL));
        assertEquals(0, progress.getScience().getQuantity(ScienceType.TABLET));
        assertEquals(1, progress.getScience().getJokers());
    }

}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/data/serializers/ScienceProgressSerializerTest.kt
package org.luxons.sevenwonders.game.data.serializers

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.luxons.sevenwonders.game.boards.ScienceType
import org.luxons.sevenwonders.game.effects.ScienceProgress
import org.luxons.sevenwonders.game.test.createScienceProgress

class ScienceProgressSerializerTest {

    private var gson: Gson? = null

    @Before
    fun setUp() {
        gson = GsonBuilder().registerTypeAdapter(ScienceProgress::class.java, ScienceProgressSerializer()).create()
    }

    @Test
    fun serialize_emptyToNull() {
        val progress = createScienceProgress(0, 0, 0, 0)
        val json = gson!!.toJson(progress)
        assertEquals("null", json)
    }

    @Test
    fun serialize_oneCompass() {
        val progress = createScienceProgress(1, 0, 0, 0)
        val json = gson!!.toJson(progress)
        assertEquals(COMPASS_STR, json)
    }

    @Test
    fun serialize_oneWheel() {
        val progress = createScienceProgress(0, 1, 0, 0)
        val json = gson!!.toJson(progress)
        assertEquals(WHEEL_STR, json)
    }

    @Test
    fun serialize_oneTablet() {
        val progress = createScienceProgress(0, 0, 1, 0)
        val json = gson!!.toJson(progress)
        assertEquals(TABLET_STR, json)
    }

    @Test
    fun serialize_oneJoker() {
        val progress = createScienceProgress(0, 0, 0, 1)
        val json = gson!!.toJson(progress)
        assertEquals(JOKER_STR, json)
    }

    @Test(expected = UnsupportedOperationException::class)
    fun serialize_failOnMultipleCompasses() {
        val progress = createScienceProgress(2, 0, 0, 0)
        gson!!.toJson(progress)
    }

    @Test(expected = UnsupportedOperationException::class)
    fun serialize_failOnMultipleWheels() {
        val progress = createScienceProgress(0, 2, 0, 0)
        gson!!.toJson(progress)
    }

    @Test(expected = UnsupportedOperationException::class)
    fun serialize_failOnMultipleTablets() {
        val progress = createScienceProgress(0, 0, 2, 0)
        gson!!.toJson(progress)
    }

    @Test(expected = UnsupportedOperationException::class)
    fun serialize_failOnMultipleJokers() {
        val progress = createScienceProgress(0, 0, 0, 2)
        gson!!.toJson(progress)
    }

    @Test(expected = UnsupportedOperationException::class)
    fun serialize_failOnMixedElements() {
        val progress = createScienceProgress(1, 1, 0, 0)
        gson!!.toJson(progress)
    }

    @Test(expected = IllegalArgumentException::class)
    fun deserialize_failOnEmptyString() {
        gson!!.fromJson("\"\"", ScienceProgress::class.java)
    }

    @Test(expected = IllegalArgumentException::class)
    fun deserialize_failOnGarbageString() {
        gson!!.fromJson("thisisgarbage", ScienceProgress::class.java)
    }

    @Test
    fun deserialize_compass() {
        val progress = gson!!.fromJson(COMPASS_STR, ScienceProgress::class.java)
        assertNotNull(progress.science)
        assertEquals(1, progress.science.getQuantity(ScienceType.COMPASS).toLong())
        assertEquals(0, progress.science.getQuantity(ScienceType.WHEEL).toLong())
        assertEquals(0, progress.science.getQuantity(ScienceType.TABLET).toLong())
        assertEquals(0, progress.science.jokers.toLong())
    }

    @Test
    fun deserialize_wheel() {
        val progress = gson!!.fromJson(WHEEL_STR, ScienceProgress::class.java)
        assertNotNull(progress.science)
        assertEquals(0, progress.science.getQuantity(ScienceType.COMPASS).toLong())
        assertEquals(1, progress.science.getQuantity(ScienceType.WHEEL).toLong())
        assertEquals(0, progress.science.getQuantity(ScienceType.TABLET).toLong())
        assertEquals(0, progress.science.jokers.toLong())
    }

    @Test
    fun deserialize_tablet() {
        val progress = gson!!.fromJson(TABLET_STR, ScienceProgress::class.java)
        assertNotNull(progress.science)
        assertEquals(0, progress.science.getQuantity(ScienceType.COMPASS).toLong())
        assertEquals(0, progress.science.getQuantity(ScienceType.WHEEL).toLong())
        assertEquals(1, progress.science.getQuantity(ScienceType.TABLET).toLong())
        assertEquals(0, progress.science.jokers.toLong())
    }

    @Test
    fun deserialize_joker() {
        val progress = gson!!.fromJson(JOKER_STR, ScienceProgress::class.java)
        assertNotNull(progress.science)
        assertEquals(0, progress.science.getQuantity(ScienceType.COMPASS).toLong())
        assertEquals(0, progress.science.getQuantity(ScienceType.WHEEL).toLong())
        assertEquals(0, progress.science.getQuantity(ScienceType.TABLET).toLong())
        assertEquals(1, progress.science.jokers.toLong())
    }

    companion object {

        private val COMPASS_STR = "\"COMPASS\""

        private val WHEEL_STR = "\"WHEEL\""

        private val TABLET_STR = "\"TABLET\""

        private val JOKER_STR = "\"any\""
    }
}
